timelock-impl/src/test/java/com/palantir/atlasdb/timelock/adjudicate/LeaderElectionDurationAccumulatorTest.java

```diff
@@ -0,0 +1,155 @@
+
+    @Test
+    public void consumeOnceForEachLeaderPair() {
+        leaderElectionResultsWithDurationInRandomOrder(LEADER_1, LEADER_2, 5, 10);
+        leaderElectionResultsWithDurationInRandomOrder(LEADER_2, LEADER_3, 5, 25);
+        leaderElectionResultsWithDurationInRandomOrder(LEADER_1, LEADER_2, 5, 1);
+        leaderElectionResultsWithDurationInRandomOrder(LEADER_1, LEADER_3, 5, 37);
+
+        verify(mockConsumer).accept(10L);
+        verify(mockConsumer).accept(25L);
+        verify(mockConsumer).accept(37L);
+        verifyNoMoreInteractions(mockConsumer);
+    }
+
+    @Test
+    @SuppressWarnings("ExecutorSubmitRunnableFutureIgnored")
+    public void testInterleavingNoPause() throws InterruptedException {
+        ExecutorService executorService = PTExecutors.newFixedThreadPool(3);
+        accumulator = new LeaderElectionDurationAccumulator(mockConsumer, 300);
+        executorService.submit(() -> leaderElectionResultsWithDurationInRandomOrder(LEADER_1, LEADER_2, 300, 1));
+        executorService.submit(() -> leaderElectionResultsWithDurationInRandomOrder(LEADER_2, LEADER_3, 300, 1017));
+        executorService.submit(() -> leaderElectionResultsWithDurationInRandomOrder(LEADER_1, LEADER_3, 300, 2076));
+        executorService.shutdown();
+        executorService.awaitTermination(1, TimeUnit.MINUTES);
+
+        verify(mockConsumer).accept(1L);
+        verify(mockConsumer).accept(1017L);
+        verify(mockConsumer).accept(2076L);
+        verifyNoMoreInteractions(mockConsumer);
+    }
+
+    @Test
+    @SuppressWarnings("ExecutorSubmitRunnableFutureIgnored")
+    public void testInterleavingWithPause() throws InterruptedException {
+        ExecutorService executorService = PTExecutors.newFixedThreadPool(3);
+        accumulator = new LeaderElectionDurationAccumulator(mockConsumer, 300);
+        executorService.submit(() -> leaderElectionResultsWithPause(LEADER_1, LEADER_2, 300, 1));
+        executorService.submit(() -> leaderElectionResultsWithPause(LEADER_2, LEADER_3, 300, 1017));
+        executorService.submit(() -> leaderElectionResultsWithPause(LEADER_1, LEADER_3, 300, 2076));
+        executorService.shutdown();
+        executorService.awaitTermination(1, TimeUnit.MINUTES);
+
+        verify(mockConsumer).accept(1L);
+        verify(mockConsumer).accept(1017L);
+        verify(mockConsumer).accept(2076L);
+        verifyNoMoreInteractions(mockConsumer);
+    }
+
+    @Test
+    @SuppressWarnings("ExecutorSubmitRunnableFutureIgnored")
+    public void testManyUpdatesForSameLeaders() throws InterruptedException {
+        ExecutorService executorService = PTExecutors.newFixedThreadPool(50);
+        int numBuckets = 100;
+        int requestsPerBucket = 50;
+        accumulator = new LeaderElectionDurationAccumulator(mockConsumer, numBuckets * requestsPerBucket);
+        List<Integer> durationBuckets =
+                IntStream.range(0, numBuckets).map(x -> x * numBuckets).boxed().collect(Collectors.toList());
```

Comment:
nit: I realise it's the same as the original test, though really the second multiplicand seems like it shouldn't be tied to `numBuckets` but `requestsPerBucket` (and I guess if you want a bit of additional space that's fine).

Reply:
Yes, I actually realised this when I wrote this. I suppose this is wrong if requestsPerBucket > numBuckets (which we could do), but I was thinking of having additional space (i.e. making sure to test with space). I think I'll just make this requestsPerBucket though, as it makes the logic easier.